aiohttp/web_jobs.py

```diff
@@ -0,0 +1,223 @@
+            with async_timeout.timeout(timeout=self._runner._timeout,
+                                       loop=self._loop):
+                yield from self._task
+        except asyncio.CancelledError:
+            pass
+        except asyncio.TimeoutError as exc:
+            context = {'message': "Job closing timed out",
+                       'job': self,
+                       'exception': exc}
+            if self._source_traceback is not None:
+                context['source_traceback'] = self._source_traceback
+            runner.call_exception_handler(context)
+
+    def _done_callback(self, task):
+        runner = self._runner
+        runner._done(self, False)
+        try:
+            exc = task.exception()
+        except asyncio.CancelledError:
+            pass
+        else:
+            if exc is not None and not self._explicit_wait:
+                context = {'message': "Job processing failed",
+                           'job': self,
+                           'exception': exc}
+                if self._source_traceback is not None:
+                    context['source_traceback'] = self._source_traceback
+                runner.call_exception_handler(context)
+                runner._failed_tasks.put_nowait(task)
+        self._runner = None  # drop backref
+        self._closed = True
+
+
+class JobRunner(Container):
+    def __init__(self, *, loop):
+        self._loop = loop
+        self._jobs = set()
+        self._timeout = 0.1
+        self._concurrency = 100
+        self._exception_handler = None
+        self._failed_tasks = asyncio.Queue(loop=loop)
+        self._failed_waiter = ensure_future(self._wait_failed(), loop=loop)
+        self._pending = deque()
+        self._active = 0
+        self._closed = False
+
+    @asyncio.coroutine
+    def run(self, coro):
+        job = Job(coro, self, self._loop)
+        self._jobs.add(job)
+        if self._active < self._concurrency:
+            self._active += 1
+            job._start()
+        else:
+            self._pending.append(job)
+        return job
+
+    def __iter__(self):
+        return iter(list(self._jobs))
+
+    def __len__(self):
+        return len(self._jobs)
+
+    def __contains__(self, job):
+        return job in self._jobs
+
+    def __repr__(self):
+        info = []
+        if self._closed:
+            info.append('closed')
+        info = ' '.join(info)
+        if info:
+            info += ' '
+        return '<JobRunner {}jobs={}>'.format(info, len(self))
+
+    @property
+    def closed(self):
+        return self._closed
+
+    @asyncio.coroutine
+    def close(self):
+        if self._closed:
+            return
+        jobs = self._jobs
+        if jobs:
+            yield from asyncio.gather(*[job.close() for job in jobs],
+                                      loop=self._loop, return_exceptions=True)
+        self._failed_tasks.put_nowait(None)
+        yield from self._failed_waiter
+        self._closed = True
```

Comment:
Pending tasks are closed without starting its coroutines.
I not sure if we need a warning here.
Because stopping web server most likely raises this warning always.